exchanges/huobi/huobi_websocket.go

```diff
@@ -75,427 +73,240 @@ var subscriptionNames = map[string]string{
-		}
+	return nil
+}
 
-		var p currency.Pair
-		var a asset.Item
-		p, a, err = h.GetRequestFormattedPairAndAssetType(data[1])
-		if err != nil {
-			return err
-		}
+func (h *HUOBI) wsHandleV2ping(_ string, respRaw []byte) error {
+	ts, err := jsonparser.GetInt(respRaw, "data", "ts")
+	if err != nil {
+		return fmt.Errorf("error getting ts from auth ping: %w", err)
+	}
+	if err := h.Websocket.AuthConn.SendJSONMessage(context.Background(), request.Unset, json.RawMessage(`{"action":"pong","data":{"ts":`+strconv.Itoa(int(ts))+`}}`)); err != nil {
+		return fmt.Errorf("error sending auth pong response: %w", err)
+	}
+	return nil
+}
 
-		h.Websocket.DataHandler <- &ticker.Price{
-			ExchangeName: h.Name,
-			Open:         wsTicker.Tick.Open,
-			Close:        wsTicker.Tick.Close,
-			Volume:       wsTicker.Tick.Amount,
-			QuoteVolume:  wsTicker.Tick.Volume,
-			High:         wsTicker.Tick.High,
-			Low:          wsTicker.Tick.Low,
-			LastUpdated:  time.UnixMilli(wsTicker.Timestamp),
-			AssetType:    a,
-			Pair:         p,
+func (h *HUOBI) wsHandleV2subResp(action string, respRaw []byte) error {
+	if ch, err := jsonparser.GetString(respRaw, "ch"); err == nil {
+		if !h.Websocket.Match.IncomingWithData(action+":"+ch, respRaw) {
+			return fmt.Errorf("%w: %s:%s", stream.ErrNoMessageListener, action, ch)
 		}
-	default:
-		h.Websocket.DataHandler <- stream.UnhandledMessageWarning{
-			Message: h.Name + stream.UnhandledMessage + string(respRaw),
-		}
-		return nil
 	}
 	return nil
 }
 
-func (h *HUOBI) sendPingResponse(pong int64) {
-	err := h.Websocket.Conn.SendJSONMessage(context.TODO(), request.Unset, WsPong{Pong: pong})
-	if err != nil {
-		log.Errorln(log.ExchangeSys, err)
-	}
+func (h *HUOBI) wsHandleChannelMsgs(s *subscription.Subscription, respRaw []byte) error {
+	switch s.Channel {
+	case subscription.TickerChannel:
+		return h.wsHandleTickerMsg(s, respRaw)
+	case subscription.OrderbookChannel:
+		return h.wsHandleOrderbookMsg(s, respRaw)
+	case subscription.CandlesChannel:
+		return h.wsHandleCandleMsg(s, respRaw)
+	case subscription.AllTradesChannel:
+		return h.wsHandleAllTradesMsg(s, respRaw)
+	case subscription.MyAccountChannel:
+		return h.wsHandleMyAccountMsg(respRaw)
+	case subscription.MyOrdersChannel:
+		return h.wsHandleMyOrdersMsg(s, respRaw)
+	case subscription.MyTradesChannel:
+		return h.wsHandleMyTradesMsg(s, respRaw)
+	}
+	return fmt.Errorf("%w: %s", common.ErrNotYetImplemented, s.Channel)
 }
 
-// WsProcessOrderbook processes new orderbook data
-func (h *HUOBI) WsProcessOrderbook(update *WsDepth, symbol string) error {
-	pairs, err := h.GetEnabledPairs(asset.Spot)
-	if err != nil {
+func (h *HUOBI) wsHandleCandleMsg(s *subscription.Subscription, respRaw []byte) error {
+	if len(s.Pairs) != 1 {
+		return subscription.ErrNotSinglePair
+	}
+	var c WsKline
+	if err := json.Unmarshal(respRaw, &c); err != nil {
 		return err
 	}
+	h.Websocket.DataHandler <- stream.KlineData{
+		Timestamp:  time.UnixMilli(c.Timestamp),
```

Comment:
Suggestion: Might be able to change `c.Timestamp` type to `types.Time` and across these changes if it can. Will leave that up to you if you want to do that.

Reply:
Ouch. Yeah. Can do.

Fixed https://github.com/gbjk/gocryptotrader/commit/e6f030cb1